lambdas/data-migration2/src/granulesAndFiles.ts

```diff
@@ -0,0 +1,269 @@
+import Knex from 'knex';
+
+import DynamoDbSearchQueue from '@cumulus/aws-client/DynamoDbSearchQueue';
+import { ApiFile } from '@cumulus/types/api/files';
+import {
+  CollectionPgModel,
+  getRecordCumulusId,
+  GranulePgModel,
+  PdrPgModel,
+  PostgresExecutionRecord,
+  PostgresFile,
+  PostgresGranule,
+  PostgresGranuleRecord,
+  ProviderPgModel,
+  tableNames,
+} from '@cumulus/db';
+import { envUtils } from '@cumulus/common';
+import Logger from '@cumulus/logger';
+
+import { RecordAlreadyMigrated } from '@cumulus/errors';
+import { MigrationSummary } from './types';
+
+const logger = new Logger({ sender: '@cumulus/data-migration/granules' });
+const Manager = require('@cumulus/api/models/base');
+const schemas = require('@cumulus/api/models/schemas');
+const { getBucket, getKey } = require('@cumulus/api/lib/FileUtils');
+
+export interface GranulesAndFilesMigrationSummary {
+  granulesSummary: MigrationSummary,
+  filesSummary: MigrationSummary,
+}
+
+/**
+ * Migrate granules record from Dynamo to RDS.
+ *
+ * @param {AWS.DynamoDB.DocumentClient.AttributeMap} dynamoRecord
+ *   Record from DynamoDB
+ * @param {Knex} knex - Knex client for writing to RDS database
+ * @returns {Promise<void>}
+ * @throws {RecordAlreadyMigrated} if record was already migrated
+ */
+export const migrateGranuleRecord = async (
+  dynamoRecord: AWS.DynamoDB.DocumentClient.AttributeMap,
+  knex: Knex
+): Promise<void> => {
+  // Validate record before processing using API model schema
+  Manager.recordIsValid(dynamoRecord, schemas.granule);
+  const [name, version] = dynamoRecord.collectionId.split('___');
+  const collectionPgModel = new CollectionPgModel();
+  const providerPgModel = new ProviderPgModel();
+  const pdrPgModel = new PdrPgModel();
+
+  const collectionCumulusId = await collectionPgModel.getRecordCumulusId(
+    knex,
+    { name, version }
+  );
+
+  const providerCumulusId = dynamoRecord.provider ? await providerPgModel.getRecordCumulusId(
+    knex,
+    { name: dynamoRecord.provider }
+  ) : undefined;
+
+  const pdrCumulusId = dynamoRecord.pdrName
+    ? await pdrPgModel.getRecordCumulusId(
+      knex,
+      { name: dynamoRecord.pdrName }
+    )
+    : undefined;
+
+  const executionCumulusId = dynamoRecord.execution
+    ? await getRecordCumulusId<PostgresExecutionRecord>(
+      { arn: dynamoRecord.execution },
+      tableNames.executions,
+      knex
+    )
+    : undefined;
+
+  const existingRecord = await knex<PostgresGranuleRecord>('granules')
+    .where({ granule_id: dynamoRecord.granuleId, collection_cumulus_id: collectionCumulusId })
```

Comment:
Not a selfish question at all! A `granuleId` isn't unique across files because many files can be associated to a single `granuleId`, so you are correct in that assumption. Though, I am not sure how I am not meeting your suggestion, does `where` not work like I thought? I am looking up a granule by `granuleId` and `collectionCumuludId`. Am I misinterpreting your suggestion? @npauzenga